File: src/evolveguard-cli/formatters.ts
import type { Baseline, EvolveGuardReport } from '../evolveguard/types';

const VERSION = '0.1.0';

export function formatRecordResult(baseline: Baseline, baselinePath: string): string {
  const lines: string[] = [];
  lines.push(`EvolveGuard v${VERSION} -- Baseline Recorded`);
  lines.push(`skill: ${baseline.skillName}  fixtures: ${baseline.fixtures.length}`);
  lines.push('');
  for (const fixture of baseline.fixtures) {
    const toolNames =
      fixture.toolCallSequence.map((e) => e.tool).join(', ') ||
      '(no capabilities detected)';
    lines.push(`  recorded  fixture: "${fixture.prompt}"  tools: ${toolNames}`);
  }
  lines.push('');
  lines.push(`baseline written to ${baselinePath}`);
  return lines.join('\n');
}

export function formatCheckResult(
  report: EvolveGuardReport,
  baselineRecordedAt: string
): string {
  const lines: string[] = [];
  const date = baselineRecordedAt.slice(0, 10);
  lines.push(`EvolveGuard v${VERSION} -- Regression Check`);
  lines.push(
    `skill: ${report.skillName}  baseline: ${date}  fixtures: ${report.summary.total}`
  );
  lines.push('');

  for (const result of report.results) {
    const tag = result.verdict === 'PASS' ? '[PASS] ' : '[DRIFT]';
    if (result.verdict === 'PASS') {
      lines.push(`${tag} fixture: "${result.prompt}"  tool-call sequence unchanged`);
    } else {
      const first = result.changes[0];
      lines.push(
        `${tag} fixture: "${result.prompt}"  ${first ? summarizeChange(first) : 'behavior changed'}`
      );
      for (const change of result.changes) {
        lines.push(`         -> ${change.message}`);
      }
    }
  }

  if (report.surfaceChanges.length > 0) {
    lines.push('');
    lines.push('skill-level surface changes (not tied to a specific fixture):');
    for (const change of report.surfaceChanges) {
      lines.push(`  -> ${change.message}`);
    }
  }

  lines.push('');
  lines.push(`${report.summary.pass} PASS, ${report.summary.drift} DRIFT, 0 FAIL`);
  if (report.exitCode === 1) {
    lines.push(
      'exit code 1 (DRIFT blocks merge by default; override with --allow-drift)'
    );
  } else {
    lines.push('exit code 0');
  }

  return lines.join('\n');
}

function summarizeChange(
  change: EvolveGuardReport['results'][number]['changes'][number]
): string {
  if (change.kind === 'added') return `new tool call: ${change.tool} (baseline had none)`;
  if (change.kind === 'removed') return `tool call removed: ${change.tool}`;
  return `scope changed: ${change.tool}`;
}

export function formatReport(report: EvolveGuardReport): string {
  const lines: string[] = [];
  lines.push(`EvolveGuard v${VERSION} -- Report`);
  lines.push(`skill: ${report.skillName}  checked: ${report.checkedAt}`);
  lines.push('');
  for (const result of report.results) {
    lines.push(`[${result.verdict}] ${result.id}: ${result.prompt}`);
    for (const change of result.changes) {
      lines.push(`  -> ${change.message}`);
    }
  }
  if (report.surfaceChanges.length > 0) {
    lines.push('');
    lines.push('skill-level surface changes:');
    for (const change of report.surfaceChanges) {
      lines.push(`  -> ${change.message}`);
    }
  }
  lines.push('');
  lines.push(`${report.summary.pass} PASS, ${report.summary.drift} DRIFT, 0 FAIL`);
  lines.push(`exit code ${report.exitCode}`);
  return lines.join('\n');
}

File: python/src/evolveguard/formatters.py
"""
Human-readable output formatting for the CLI. Ported from
src/evolveguard-cli/formatters.ts.
"""
from __future__ import annotations

from typing import List

from . import __version__ as _VERSION
from .types import Baseline, CapabilityChange, EvolveGuardReport


def format_record_result(baseline: Baseline, baseline_path: str) -> str:
    lines: List[str] = []
    lines.append(f"EvolveGuard v{_VERSION} -- Baseline Recorded")
    lines.append(f"skill: {baseline.skill_name}  fixtures: {len(baseline.fixtures)}")
    lines.append("")
    for fixture in baseline.fixtures:
        tool_names = ", ".join(e.tool for e in fixture.tool_call_sequence) or (
            "(no capabilities detected)"
        )
        lines.append(f'  recorded  fixture: "{fixture.prompt}"  tools: {tool_names}')
    lines.append("")
    lines.append(f"baseline written to {baseline_path}")
    return "\n".join(lines)


def _summarize_change(change: CapabilityChange) -> str:
    if change.kind == "added":
        return f"new tool call: {change.tool} (baseline had none)"
    if change.kind == "removed":
        return f"tool call removed: {change.tool}"
    return f"scope changed: {change.tool}"


def format_check_result(report: EvolveGuardReport, baseline_recorded_at: str) -> str:
    lines: List[str] = []
    date = baseline_recorded_at[:10]
    lines.append(f"EvolveGuard v{_VERSION} -- Regression Check")
    lines.append(
        f"skill: {report.skill_name}  baseline: {date}  fixtures: {report.summary.total}"
    )
    lines.append("")

    for result in report.results:
        tag = "[PASS] " if result.verdict == "PASS" else "[DRIFT]"
        if result.verdict == "PASS":
            lines.append(f'{tag} fixture: "{result.prompt}"  tool-call sequence unchanged')
        else:
            first = result.changes[0] if result.changes else None
            summary = _summarize_change(first) if first else "behavior changed"
            lines.append(f'{tag} fixture: "{result.prompt}"  {summary}')
            for change in result.changes:
                lines.append(f"         -> {change.message}")

    if report.surface_changes:
        lines.append("")
        lines.append("skill-level surface changes (not tied to a specific fixture):")
        for change in report.surface_changes:
            lines.append(f"  -> {change.message}")

    lines.append("")
    lines.append(f"{report.summary.pass_count} PASS, {report.summary.drift} DRIFT, 0 FAIL")
    if report.exit_code == 1:
        lines.append("exit code 1 (DRIFT blocks merge by default; override with --allow-drift)")
    else:
        lines.append("exit code 0")

    return "\n".join(lines)


def format_report(report: EvolveGuardReport) -> str:
    lines: List[str] = []
    lines.append(f"EvolveGuard v{_VERSION} -- Report")
    lines.append(f"skill: {report.skill_name}  checked: {report.checked_at}")
    lines.append("")
    for result in report.results:
        lines.append(f"[{result.verdict}] {result.id}: {result.prompt}")
        for change in result.changes:
            lines.append(f"  -> {change.message}")
    if report.surface_changes:
        lines.append("")
        lines.append("skill-level surface changes:")
        for change in report.surface_changes:
            lines.append(f"  -> {change.message}")
    lines.append("")
    lines.append(f"{report.summary.pass_count} PASS, {report.summary.drift} DRIFT, 0 FAIL")
    lines.append(f"exit code {report.exit_code}")
    return "\n".join(lines)
